datajunction-server/datajunction_server/construction/build_v2.py

```diff
@@ -0,0 +1,938 @@
+    apply_filters_to_node(node, query, to_filter_asts(filters))
+
+    query.ctes.extend(new_cte_mapping.values())
+    query.select.add_aliases_to_unnamed_columns()
+    return query
+
+
+def apply_filters_to_node(
+    node: NodeRevision,
+    query: ast.Query,
+    filters: List[ast.Expression],
+):
+    """
+    Apply pushdown filters if possible to the node's query AST.
+
+    A pushdown filter is defined as a filter with references to dimensions that
+    already exist on the node query without any additional dimension joins. We can
+    apply these filters directly to the query AST by renaming the dimension ref in
+    the filter expression.
+    """
+    for filter_ast in filters:
+        all_referenced_dimensions_can_pushdown = True
+        for filter_dim in filter_ast.find_all(ast.Column):
+            column_name = get_column_from_canonical_dimension(
+                filter_dim.identifier(),
+                node,
+            )
+            node_col = (
+                query.select.column_mapping.get(column_name) if column_name else None
+            )
+            if node_col:
+                replacement = (
+                    node_col.child if isinstance(node_col, ast.Alias) else node_col  # type: ignore
+                )
+                filter_ast.replace(filter_dim, replacement)
+            else:
+                all_referenced_dimensions_can_pushdown = False
+
+        if all_referenced_dimensions_can_pushdown:
+            query.select.where = combine_filter_conditions(
+                query.select.where,
+                filter_ast,
+            )
+    return query
+
+
+def get_dj_node_references_from_select(
+    select: ast.SelectExpression,
+) -> DefaultDict[NodeRevision, List[ast.Table]]:
+    """
+    Extract all DJ node references (source, transform, dimensions) from the select
+    expression. DJ node references are represented in the AST as table expressions
+    and have an attached DJ node.
+    """
+    tables: DefaultDict[NodeRevision, List[ast.Table]] = collections.defaultdict(list)
+
+    for table in select.find_all(ast.Table):
+        if node := table.dj_node:  # pragma: no cover
+            tables[node].append(table)
+    return tables
+
+
+def get_table_for_node(
+    node: NodeRevision,
+    build_criteria: Optional[BuildCriteria] = None,
+) -> Optional[Union[ast.Select, ast.Table]]:
+    """
+    If a node has a materialized table available, return the materialized table.
+    Source nodes should always have an associated table, whereas for all other nodes
+    we can check the materialization type.
+    """
+    table = None
+    can_use_materialization = (
+        build_criteria and node.name != build_criteria.target_node_name
+    )
+    if node.type == NodeType.SOURCE:
+        table_name = (
+            f"{node.catalog.name}.{node.schema_}.{node.table}"
+            if node.schema_ == "iceberg"
+            else f"{node.schema_}.{node.table}"
+        )
```

Comment:
I didn't know we store this `schema_` value on nodes. Is it true that only iceberg tables require the catalog name? Wouldn't hive tables or any table queried via Trino also need the catalog name included?

Reply:
We only store `schema_` for source nodes, as far as I can tell. And yes, I recall that there's a weird edge case where if the schema is `iceberg`, it needs the catalog name or the Spark query will fail to run (this is copied from our existing build fyi).